src/librustc_traits/type_op_prove_predicate.rs

```diff
@@ -8,25 +8,31 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use rustc::infer::InferOk;
 use rustc::infer::canonical::{Canonical, QueryResult};
 use rustc::traits::query::type_op::prove_predicate::ProvePredicate;
 use rustc::traits::query::NoSolution;
-use rustc::traits::{FulfillmentContext, Obligation, ObligationCause, TraitEngine};
+use rustc::traits::{Obligation, ObligationCause};
 use rustc::ty::TyCtxt;
 use rustc_data_structures::sync::Lrc;
-use syntax::codemap::DUMMY_SP;
 
 crate fn type_op_prove_predicate<'tcx>(
     tcx: TyCtxt<'_, 'tcx, 'tcx>,
     canonicalized: Canonical<'tcx, ProvePredicate<'tcx>>,
 ) -> Result<Lrc<Canonical<'tcx, QueryResult<'tcx, ()>>>, NoSolution> {
-    let tcx = tcx.global_tcx();
-    tcx.infer_ctxt().enter(|ref infcx| {
-        let (ProvePredicate { param_env, predicate }, canonical_inference_vars) =
-            infcx.instantiate_canonical_with_fresh_inference_vars(DUMMY_SP, &canonicalized);
-        let fulfill_cx = &mut FulfillmentContext::new();
-        let obligation = Obligation::new(ObligationCause::dummy(), param_env, predicate);
-        fulfill_cx.register_predicate_obligation(infcx, obligation);
-        infcx.make_canonicalized_query_result(canonical_inference_vars, (), fulfill_cx)
-    })
+    tcx.infer_ctxt()
+        .enter_canonical_trait_query(&canonicalized, |_infcx, key| {
+            let ProvePredicate {
+                param_env,
+                predicate,
+            } = key;
+            Ok(InferOk {
+                value: (),
+                obligations: vec![Obligation::new(
+                    ObligationCause::dummy(),
+                    param_env,
+                    predicate,
+                )],
+            })
```

Comment:
Interesting, I would've expected you to pass `fulfill_cx` to the closure instead.